File: backend/app/schemas/communication.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID
from pydantic import BaseModel, Field, ConfigDict
from app.models.communication import SmsProviderType, WhatsAppProviderType


# ── Preferences Schemas ────────────────────────────────────────────────────────

class UserCommunicationPreferenceResponse(BaseModel):
    id: UUID
    school_id: UUID
    user_id: UUID
    enable_in_app: bool
    enable_email: bool
    enable_sms: bool
    enable_whatsapp: bool
    enable_attendance: bool
    enable_fees: bool
    enable_exams: bool
    enable_events: bool
    enable_hostel: bool
    enable_leave: bool
    enable_emergency: bool
    enable_announcements: bool

    model_config = ConfigDict(from_attributes=True)


class UserCommunicationPreferenceUpdate(BaseModel):
    enable_in_app: bool | None = None
    enable_email: bool | None = None
    enable_sms: bool | None = None
    enable_whatsapp: bool | None = None
    enable_attendance: bool | None = None
    enable_fees: bool | None = None
    enable_exams: bool | None = None
    enable_events: bool | None = None
    enable_hostel: bool | None = None
    enable_leave: bool | None = None
    enable_announcements: bool | None = None


# ── School Communication Config Schemas ──────────────────────────────────────

class SchoolCommunicationConfigResponse(BaseModel):
    id: UUID
    school_id: UUID
    sms_provider: SmsProviderType
    whatsapp_provider: WhatsAppProviderType
    sms_enabled: bool
    whatsapp_enabled: bool
    sms_sender_id: str | None = None
    sms_entity_id: str | None = None
    whatsapp_phone_number_id: str | None = None
    whatsapp_business_account_id: str | None = None
    sms_configured: bool = False
    whatsapp_configured: bool = False
    sms_api_key_masked: str | None = None
    whatsapp_access_token_masked: str | None = None
    sms_monthly_quota: int = 10000
    sms_sent_this_month: int = 0
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class SchoolCommunicationConfigUpdate(BaseModel):
    sms_provider: SmsProviderType | None = None
    whatsapp_provider: WhatsAppProviderType | None = None
    sms_enabled: bool | None = None
    whatsapp_enabled: bool | None = None
    sms_api_key: str | None = None
    sms_sender_id: str | None = None
    sms_entity_id: str | None = None
    whatsapp_access_token: str | None = None
    whatsapp_phone_number_id: str | None = None
    whatsapp_business_account_id: str | None = None
    sms_monthly_quota: int | None = None


# ── Notification Template Schemas ─────────────────────────────────────────────

class NotificationTemplateCreate(BaseModel):
    template_key: str = Field(..., max_length=100)
    name: str = Field(..., max_length=150)
    category: str = Field(default="ANNOUNCEMENT")
    title_template: str = Field(..., min_length=2)
    body_template: str = Field(..., min_length=2)
    dlt_entity_id: str | None = None
    dlt_template_id: str | None = None
    whatsapp_template_name: str | None = None
    whatsapp_language_code: str | None = None


class NotificationTemplateResponse(BaseModel):
    id: UUID | str
    template_key: str
    name: str
    category: str
    title_template: str
    body_template: str
    is_active: bool
    is_custom: bool = True
    dlt_entity_id: str | None = None
    dlt_template_id: str | None = None
    whatsapp_template_name: str | None = None
    whatsapp_language_code: str | None = None

    model_config = ConfigDict(from_attributes=True)



# ── Inbox & Provider Status Schemas ──────────────────────────────────────────

class InboxItemResponse(BaseModel):
    id: str
    title: str
    body: str
    template_key: str
    channel: str
    status: str
    is_read: bool
    created_at: str | None = None


class ProviderStatusResponse(BaseModel):
    channel: str
    provider_name: str
    is_configured: bool
    status: str


# ── Notification Detail & Analytics Schemas ──────────────────────────────────

class NotificationDetailResponse(BaseModel):
    id: UUID
    school_id: UUID
    recipient_type: str
    recipient_id: UUID | None = None
    recipient_name: str
    recipient_contact: str
    channel: str
    template_key: str
    title: str
    body: str
    status: str
    error_message: str | None = None
    sent_at: datetime | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None
    provider_name: str | None = None
    provider_message_id: str | None = None
    retry_count: int = 0
    max_retries: int = 3
    idempotency_key: str | None = None

    model_config = ConfigDict(from_attributes=True)


class DailyVolumeItem(BaseModel):
    date: str
    count: int
    sent: int = 0
    failed: int = 0


class NotificationAnalyticsResponse(BaseModel):
    total_notifications: int
    sent_count: int
    delivered_count: int
    failed_count: int
    pending_count: int
    cancelled_count: int
    success_rate_percent: float
    failure_rate_percent: float
    pending_rate_percent: float
    by_channel: dict[str, int]
    by_event: dict[str, int]
    daily_volume: list[DailyVolumeItem]
    providers: list[ProviderStatusResponse]

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/api/communication.ts
import { apiClient } from '@/services/api/client';

export interface UserPreferences {
  id: string;
  school_id: string;
  user_id: string;
  enable_in_app: boolean;
  enable_email: boolean;
  enable_sms: boolean;
  enable_whatsapp: boolean;
  enable_attendance: boolean;
  enable_fees: boolean;
  enable_exams: boolean;
  enable_events: boolean;
  enable_hostel: boolean;
  enable_leave: boolean;
  enable_emergency: boolean;
  enable_announcements: boolean;
}

export interface NotificationItem {
  id: string;
  recipient_type?: string;
  recipient_id?: string | null;
  recipient_name?: string;
  recipient_contact?: string;
  channel: string;
  template_key: string;
  title: string;
  body: string;
  status: string;
  error_message?: string;
  retry_count?: number;
  max_retries?: number;
  is_read?: boolean;
  provider_name?: string | null;
  provider_message_id?: string | null;
  idempotency_key?: string | null;
  sent_at?: string;
  created_at?: string;
  updated_at?: string;
}

export interface NotificationDetail extends NotificationItem {
  school_id: string;
}

export interface NotificationTemplate {
  id: string;
  template_key: string;
  name: string;
  category: string;
  title_template: string;
  body_template: string;
  is_active: boolean;
  is_custom: boolean;
  dlt_entity_id?: string | null;
  dlt_template_id?: string | null;
  whatsapp_template_name?: string | null;
  whatsapp_language_code?: string | null;
}

export interface SchoolCommunicationConfig {
  id: string;
  school_id: string;
  sms_provider: 'NONE' | 'FAST2SMS' | 'TWILIO' | 'MOCK';
  whatsapp_provider: 'NONE' | 'META_WHATSAPP_CLOUD' | 'TWILIO_WHATSAPP' | 'MOCK';
  sms_enabled: boolean;
  whatsapp_enabled: boolean;
  sms_sender_id?: string | null;
  sms_entity_id?: string | null;
  whatsapp_phone_number_id?: string | null;
  whatsapp_business_account_id?: string | null;
  sms_configured: boolean;
  whatsapp_configured: boolean;
  sms_api_key_masked?: string | null;
  whatsapp_access_token_masked?: string | null;
  sms_monthly_quota: number;
  sms_sent_this_month: number;
  created_at: string;
  updated_at: string;
}

export interface SchoolCommunicationConfigUpdate {
  sms_provider?: 'NONE' | 'FAST2SMS' | 'TWILIO' | 'MOCK';
  whatsapp_provider?: 'NONE' | 'META_WHATSAPP_CLOUD' | 'TWILIO_WHATSAPP' | 'MOCK';
  sms_enabled?: boolean;
  whatsapp_enabled?: boolean;
  sms_api_key?: string | null;
  sms_sender_id?: string | null;
  sms_entity_id?: string | null;
  whatsapp_access_token?: string | null;
  whatsapp_phone_number_id?: string | null;
  whatsapp_business_account_id?: string | null;
  sms_monthly_quota?: number;
}

export interface ProviderStatus {
  channel: string;
  provider_name: string;
  is_configured: boolean;
  status: string;
}

export interface DailyVolumeItem {
  date: string;
  count: number;
  sent: number;
  failed: number;
}

export interface DeliveryMetrics {
  total_notifications: number;
  sent_count: number;
  failed_count: number;
  pending_count: number;
  cancelled_count: number;
  failure_rate_percent: number;
  by_channel: Record<string, number>;
  providers: ProviderStatus[];
}

export interface NotificationAnalytics {
  total_notifications: number;
  sent_count: number;
  delivered_count: number;
  failed_count: number;
  pending_count: number;
  cancelled_count: number;
  success_rate_percent: number;
  failure_rate_percent: number;
  pending_rate_percent: number;
  by_channel: Record<string, number>;
  by_event: Record<string, number>;
  daily_volume: DailyVolumeItem[];
  providers: ProviderStatus[];
}

export interface NotificationFilterParams {
  page?: number;
  page_size?: number;
  status?: string;
  channel?: string;
  event_type?: string;
  start_date?: string;
  end_date?: string;
  search?: string;
}

export const communicationApi = {
  fetchUserInbox: async (page = 1, page_size = 20) => {
    const res = await apiClient.get('/notifications/inbox', { params: { page, page_size } });
    return res.data.data;
  },

  fetchUnreadCount: async () => {
    const res = await apiClient.get('/notifications/unread-count');
    return res.data.data.unread_count as number;
  },

  markNotificationRead: async (id: string) => {
    const res = await apiClient.post(`/notifications/inbox/${id}/read`);
    return res.data;
  },

  markAllNotificationsRead: async () => {
    const res = await apiClient.post('/notifications/inbox/read-all');
    return res.data;
  },

  fetchUserPreferences: async () => {
    const res = await apiClient.get('/notifications/preferences');
    return res.data.data as UserPreferences;
  },

  updateUserPreferences: async (payload: Partial<UserPreferences>) => {
    const res = await apiClient.put('/notifications/preferences', payload);
    return res.data.data as UserPreferences;
  },

  fetchSchoolCommunicationConfig: async () => {
    const res = await apiClient.get('/notifications/config');
    return res.data.data as SchoolCommunicationConfig;
  },

  updateSchoolCommunicationConfig: async (payload: SchoolCommunicationConfigUpdate) => {
    const res = await apiClient.put('/notifications/config', payload);
    return res.data.data as SchoolCommunicationConfig;
  },

  fetchProviderStatuses: async () => {
    const res = await apiClient.get('/notifications/providers/status');
    return res.data.data as ProviderStatus[];
  },

  fetchDeliveryLogs: async (params?: NotificationFilterParams) => {
    const res = await apiClient.get('/notifications', { params });
    return res.data.data;
  },

  fetchNotificationDetail: async (id: string) => {
    const res = await apiClient.get(`/notifications/${id}`);
    return res.data.data as NotificationDetail;
  },

  fetchDeliveryMetrics: async () => {
    const res = await apiClient.get('/notifications/delivery-metrics');
    return res.data.data as DeliveryMetrics;
  },

  fetchNotificationAnalytics: async (params?: { start_date?: string; end_date?: string }) => {
    const res = await apiClient.get('/notifications/analytics', { params });
    return res.data.data as NotificationAnalytics;
  },

  retryNotification: async (id: string) => {
    const res = await apiClient.post(`/notifications/${id}/retry`);
    return res.data.data;
  },

  fetchTemplates: async () => {
    const res = await apiClient.get('/notifications/templates');
    return res.data.data as NotificationTemplate[];
  },

  createTemplate: async (payload: {
    template_key: string;
    name: string;
    category: string;
    title_template: string;
    body_template: string;
    dlt_entity_id?: string;
    dlt_template_id?: string;
    whatsapp_template_name?: string;
    whatsapp_language_code?: string;
  }) => {
    const res = await apiClient.post('/notifications/templates', payload);
    return res.data.data;
  },

  sendAnnouncement: async (payload: { title: string; message: string; recipient_name: string; recipient_contact: string; channel: string }) => {
    const res = await apiClient.post('/notifications/send', payload);
    return res.data.data;
  },
};

